src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2221,7 +2222,21 @@ RunDistributedExecution(DistributedExecution *execution)
 		/* always (re)build the wait event set the first time */
 		execution->rebuildWaitEventSet = true;
 
-		while (execution->unfinishedTaskCount > 0 && !cancellationReceived)
+		/*
+		 * Iterate until all the tasks are finished. Once all the tasks
+		 * are finished, ensure that that all the connection initializations
+		 * are also finished. Otherwise, those connections are terminated
+		 * abruptly before they are established (or failed). Instead, we let
+		 * the ConnectionStateMachine() to properly handle them.
+		 *
+		 * Note that we could have the connections that are not established
+		 * as a side effect of slow-start algorithm. At the time the algorithm
+		 * decides to establish new connections, the execution might have tasks
+		 * to finish. But, the execution might finish before the new connections
+		 * are established.
+		 */
+		while (execution->unfinishedTaskCount > 0 ||
```

Comment:
maybe we can define a function `HasUnfinishedTask` to check `execution->unfinishedTaskCount` use it in https://github.com/citusdata/citus/pull/4317/files#diff-1b46c5ae7c0832742e4a851ace6c1f85ba23e191d71ff2e65489fbbc6ba2b2a2R2981 as well

Reply:
@onurctirtir I accidentally removed this comment, actually marked as "spam", and cannot undo it.  Do you remember what was your comment :D

Reply:
iirc, it was a minor/optional item like:

maybe we can define a one-liner function to check `execution->unfinishedTaskCount > 0` and we can use it in the [other line](https://github.com/citusdata/citus/pull/4317/files#diff-1b46c5ae7c0832742e4a851ace6c1f85ba23e191d71ff2e65489fbbc6ba2b2a2R3007) too